### ISSUE TITLE
adam optimizer does not work with FP16

### ISSUE THREAD

**Issue by husmen** [keras-team/keras#11860]
Hi,

I noticed some weird behaviour when training a sequential model using FP16, `mse` as loss function and `adam` as optimizer; the loss can't be calculated and I get an output similar to:

    Epoch 1/10
    1381/1381 [==============================] - 1033s 748ms/step - loss: nan

There is no issue when using FP32, or changing the optimizer with FP16 (I tried both adamax and sgd). Am I missing something or is there something wrong with the implementation of `adam`?

It is worth noting that I'm using an RTX 2080 which does support FP16.

- [ * ] Check that you are up-to-date with the master branch of Keras. You can update with:
`pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps`
    - done

- [ * ] Check that your version of TensorFlow is up-to-date. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).
    - done

- [ * ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).
    - gist link with code snippet [here](https://gist.github.com/husmen/8c9ef0eef2cfb6b5621b51f143e60b09)
    - link for the original full code (before adding FP16) [here](https://github.com/openzeka/racecar-controllers/blob/8eb48ebda8d86c8d1995c04251a695ca6e99e844/marc-examples/ktrain/model_trainer.ipynb)

**Comment by gabrieldemarmiesse** [comment]
Not all operations work correctly with FP16. This is nothing we can do about. Half the running time doesn't come for free. See http://on-demand.gputechconf.com/gtc/2017/presentation/s7218-training-with-mixed-precision-boris-ginsburg.pdf for example

**Comment by husmen** [comment]
I thought I should include the solution I found here and close the issue

my specific problem was due to the default `epsilon` value of `1e-7` beeing too small for FP16 and leading to `nan` values, changing it with `K.set_epsilon(1e-4)` solves the problem. Well, at least for mse, msle, and mae metrics, mape and cosine still give `inf` and `nan` respectively.
